File: M08_Architecture/C02_Component/S01_ViewModel/src/main/java/net/bi4vmr/study/base/MyViewModel.java
package net.bi4vmr.study.base;

import android.util.Log;

import androidx.lifecycle.ViewModel;

import java.util.UUID;

/**
 * Name        : MyViewModel
 * <p>
 * Author      : BI4VMR
 * <p>
 * Email       : bi4vmr@outlook.com
 * <p>
 * Date        : 2023-05-11 21:10
 * <p>
 * Description : ViewModel：基本功能测试。
 */
public class MyViewModel extends ViewModel {

    private static final String TAG = "TestApp-" + TestFragment.class.getSimpleName();

    // 实例ID
    public final String id;

    // 共享数据
    private int num = 0;

    public MyViewModel() {
        // 生成随机ID，标识当前实例。
        id = genRandomID();
        Log.i(TAG, "VM created. ID: " + id);
    }

    // 读取共享数据
    public int getNum() {
        return num;
    }

    // 写入共享数据
    public void setNum(int num) {
        this.num = num;
    }

    // 获取随机ID
    private String genRandomID() {
        return UUID.randomUUID()
                .toString()
                .toUpperCase()
                .substring(0, 6);
    }
}

File: M04_System/C06_MultiTask/S01_Base/src/main/kotlin/net/bi4vmr/study/base/MyViewModel.kt
package net.bi4vmr.study.base

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.coroutines.cancellation.CancellationException

/**
 * 测试ViewModel。
 *
 * @since 1.0.0
 * @author bi4vmr@outlook.com
 */
class MyViewModel : ViewModel() {

    companion object {
        private val TAG: String = "TestApp-${MyViewModel::class.java.simpleName}"
    }

    // 测试协程
    fun task() {
        viewModelScope.launch {
            try {
                Log.d(TAG, "Task start.")
                delay(5000L)
                Log.d(TAG, "Task end.")
            } catch (e: CancellationException) {
                Log.d(TAG, "Task was canceled!")
            }
        }
    }

    override fun onCleared() {
        Log.d(TAG, "OnCleared.")
    }
}
